design/design-doc-template-stacks-experience.md

```diff
@@ -0,0 +1,473 @@
+behaviors:
+  crds:
+    # This is a particular CRD which is being configured. When the
+    # controller sees an object of this type, it will do something.
+    wordpressinstance.wordpress.samples.stacks.crossplane.io/v1alpha1:
+      # These are the templates which should be rendered when an object
+      # of the type above is seen.
+      #
+      # Defaults for the variables can be defined in the default values
+      # for the CRD fields, using the standard Kubernetes mechanism for
+      # specifying CRD field default values.
+      resources:
+        # These are individual files which will be rendered.
+        - templates/kubernetescluster.yaml
+        - templates/mysqlinstance.yaml
+        - tempaltes/kubernetesapplication.yaml
+EOF
+```
+
+The `stack.yaml` should be in the build root of the repository. For most
+single-stack repositories, this means the root of the repository.
+
+### Specifying default values
+
+Default values will be specified in the CRD itself, using the [standard
+mechanism][kubernetes-default-values] for specifying default values for
+CRD fields. Here is an example excerpt from a realistic CRD, where the
+CRD's `spec.image` field is configured with a default:
+
+```yaml
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: null
+  name: wordpressinstances.wordpress.samples.stacks.crossplane.io
+spec:
+  group: wordpress.samples.stacks.crossplane.io
+  names:
+    kind: WordpressInstance
+    plural: wordpressinstances
+  scope: ""
+  validation:
+    openAPIV3Schema:
+      description: WordpressInstance is the Schema for the wordpressinstances API
+      properties:
+        ...
+        spec:
+          type: object
+          properties:
+            ...
+            image:
+              type: string
+              description: A custom wordpress container image id to use
+              # Defaults are specified like this, using the schema validation for CRD fields.
+              # For more about how this works with CRDs, see:
+              # https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#defaulting
+              default: "wordpress:4.6.1-apache"
+```
+
+For the full example, see the sample in the
+[template stack experience repository](https://github.com/suskin/template-stack-experience/tree/master/wordpress-workload/go-templates/default-variables).
+
+### How processing objects works under the covers
+
+The imagined implementation for how objects sent to the stack for
+rendering become resources is that the object's fields become the input
+for the template rendered by the controller. For more details, see the
+[detailed examples](https://github.com/suskin/template-stack-experience)
+or the [design document about the
+internals](https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-template-stacks.md).
+
+### Templating/configuration engine
+
+Template stacks will not be opinionated about which templating engine is
```

Comment:
Can we go into a lot more detail here about the experience around using existing engines? If that is the general direction we are heading in, I'd love to see what it looks like for a stack author to basically reuse their existing helm chart inside the new stack they are building.  How do they specify what engine to use at render time? How do they include their existing assets/artifacts into the stack repo/package?

Reply:
I've added the links to other sections, and I've also linked out to the recently created [quick start example](https://github.com/suskin/template-stack-experience/tree/master/wordpress-workload/quick-start) in the experience sample repository to give readers somewhere to go to get more details about this.